### PR TITLE
Notifications fixes for groups and adhoc groups

### LINKED ISSUE #7229: No "needs help" status in the exercise report page.
The exercise report page does not show the "needs help" status.

### Observed behavior

I see two (or three) issues while testing [coach-track-lesson-groups-notifications.feature - Gherkin story](https://github.com/learningequality/kolibri/blob/release-v0.14.x/integration_testing/features/coach/coach-track-lesson-groups-notifications.feature#L51)

1. I cannot see the "needs help" status in the exercise report page.  However when I checked the `View by groups`, it shows the *i*/needs help red icon but it has no values.  I expected "needs help" to replace the "started" status.
Screen Shot 2020-07-01 at 5 31 53 PM

2. The status icons are helpful in the page BUT it only shows when the `View by groups` is checked.  It should be displayed in all views (grouped or ungrouped).
Screen Shot 2020-07-01 at 5 39 20 PM

Here's the `Difficult Questions` tab showing the "needs help" status.
Screen Shot 2020-07-01 at 5 31 58 PM

Here's the `.gif` of the exercise page as I navigated it:
2020-07-01-wed--1--v0 14 0-b7--coach track lesson group notification 4

3. And perhaps we can also update the Lessons page to show the "needs help" status as another column?  I think it's very helpful to the coaches.
Screen Shot 2020-07-01 at 6 03 23 PM

### Expected behavior
The "needs help" status must be shown in the exercise report page.  I expected it to replace the "Started" status.

### User-facing consequences
As a coach, I will not see which learners "needs help".

### Errors and logs
[daemon.txt](https://github.com/learningequality/kolibri/files/4857063/daemon.txt)
[kolibri.txt](https://github.com/learningequality/kolibri/files/4857066/kolibri.txt)

### Steps to reproduce

1. Install [v0.14.0-beta7](https://github.com/learningequality/kolibri/releases/tag/v0.14.0-beta7) in a Win7 VM.
1. Follow this [coach-track-lesson-groups-notifications.feature - Gherkin story](https://github.com/learningequality/kolibri/blob/release-v0.14.x/integration_testing/features/coach/coach-track-lesson-groups-notifications.feature#L51). 

### Context

* Kolibri version: [v0.14.0-beta7](https://github.com/learningequality/kolibri/releases/tag/v0.14.0-beta7)
* Operating system: Win7 on VM
* Browser: Firefox v77
* Python: v3.6.8

### PR DESCRIPTION
**Pull request by rtibbles**
### Summary
Fixes several extant bugs in our coach notifications:
* Major fix - fix bug where 'needs help' would not show up in classroom reports if the notification was created while a learner was assigned to the exercise via a lesson assigned to a group or adhoc group
* Migrates our `classroom_id` field to only refer to the classroom in which the notification was created
* Adds an `assigment_collections` JSON Field which contains a JSON array of collection ids in which this activity was logged (classroom, groups, and ad hoc groups)
* Updates the notifications backend machinery to record notifications with this new data structure
* Runs a data migration to convert all existing notifications into this structure, populating `assignment_collections` from the current `classroom_id` field
* Fixes a bug whereby links to classroom reports would not get generated for adhoc group related notifications
* Fixes a bug whereby Answered events would not log the associated `contentnode_id`
* Simplifies the pagination logic for the notifications endpoint, by simpler returning whether there are more relevant notifications to fetch or not
* Consolidates all notifications in the frontend into a single vuex state

### Reviewer guidance
Does it fix the bug described in #7229?
Do notifications all work as expected?

One extant question - should we not be aggregating adhoc learner notifications, or should we group them together like a group?

### References
Fixes #7229 

----

### Contributor Checklist

PR process:

- [x] PR has the correct target branch and milestone
- [x] PR has 'needs review' or 'work-in-progress' label
- [x] If PR is ready for review, a reviewer has been added. (Don't use 'Assignees')
- [ ] If this is an important user-facing change, PR or related issue has a 'changelog' label
- [ ] If this includes an internal dependency change, a link to the diff is provided

Testing:

- [x] Contributor has fully tested the PR manually
- [ ] If there are any front-end changes, before/after screenshots are included
- [x] Critical user journeys are covered by Gherkin stories
- [x] Critical and brittle code paths are covered by unit tests

### Reviewer Checklist

- Automated test coverage is satisfactory
- PR is fully functional
- PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- External dependency files were updated if necessary (`yarn` and `pip`)
- Documentation is updated
- Contributor is in AUTHORS.md

### CODE REVIEW

**Comment by jonboiser** [comment]
I was able to confirm that the basic issue in #7229 is fixed. I was able to see the "Needs help" indicator for my user in all the relevant parts of the Coach page. However, I'm wondering if there is some filtering going on in the "Class activity' vs. the full Notification feed. Compare these two screenshots taken at the same time.

The "Class activity" section is missing some notifications:

CleanShot 2020-07-28 at 12 37 31@2x
CleanShot 2020-07-28 at 12 37 25@2x

**Comment by rtibbles** [comment]
I had noticed this as well, but had assumed I was failing to understand the original summarization logic - will revisit this and ensure all desired notifications are being displayed!

File: kolibri/plugins/coach/assets/src/modules/coachNotifications/getters.js [hunk 1]
```diff
@@ -99,59 +172,22 @@ export function summarizedNotifications(state, getters, rootState, rootGetters)
...
-          learnerSummary: {
-            firstUserName: firstUser.user,
-            firstUserId: firstUser.user_id,
-            total: partitioning.hasEvent.length,
-            // not used for Needs Help
-            completesCollection: partitioning.rest.length === 0,
-          },
-        });
-      }
+      summaryEvents.push({
```
Comment (nucleogenesis) [inline kolibri/plugins/coach/assets/src/modules/coachNotifications/getters.js]: :heart_eyes:

**Review by nucleogenesis — DISMISSED** [review]
I made a bunch of quizzes & lessons and took them incognito on `release-0.14.x` then tested the migration and everything worked great. Mixed groups, individual learners, etc.

Notifications come through as expected and are bundled as expected.

The only very very minor thing is that a new notification that ought to be bundled flashes visibly for a second before being bundled - I cropped the gif weirdly - so it only shows the "Initial state" with one notification for a second before it flashes and then combines the second.

notification-flash

**Comment by jonboiser** [comment]
@nucleogenesis Did you see the same thing I saw, the Home Page > Class Activity version of the feed only reflected a subset of the full notification feed?

**Comment by jonboiser** [comment]
I'm still not seeing `HelpNeeded` notifications on the home page. In the summarizedNotifications getter, I'm not seeing any `HelpNeeded` events getting grouped by this line:

https://github.com/rtibbles/kolibri/blob/c71f8ed00b12be0be5259dfc407b054ab02b6658/kolibri/plugins/coach/assets/src/modules/coachNotifications/getters.js#L145

Even though I'm seeing some of the events when I console.log the `groupedNotification` variable a couple lines above.

image

image

**Comment by jonboiser** [comment]
It might be because of this `uniq`-ing line here:

https://github.com/rtibbles/kolibri/blob/c71f8ed00b12be0be5259dfc407b054ab02b6658/kolibri/plugins/coach/assets/src/modules/coachNotifications/getters.js#L136-L137

Basically, the `Answered` notification wins if it's first in the array, since the uniqueness criteria is just the userID

File: kolibri/plugins/coach/assets/src/modules/coachNotifications/getters.js [hunk 2]
```diff
@@ -1,159 +1,185 @@
...
     if (n.object === QUIZ) {
-      return `${n.type}_${n.quiz_id}`;
+      return `${n.object}_${n.quiz_id}`;
     }
   });
 
   for (let groupCode in groupedNotifications) {
-    const allEvents = groupedNotifications[groupCode];
+    // Filter out all bust the most recent event for each user
+    const allEvents = sortedUniqBy(groupedNotifications[groupCode], 'user_id');
```
Comment (jonboiser) [inline kolibri/plugins/coach/assets/src/modules/coachNotifications/getters.js:142]: To fix the issue I've had with missing `NeedsHelp` notifications, I filter `groupedNotifications[groupCode]` with a predicate `({event}) => event !== NotificationEvents.ANSWERED`.

Comment (jonboiser) [reply kolibri/plugins/coach/assets/src/modules/coachNotifications/getters.js:142]: Screenshot after filtering out `Answered` events.

CleanShot 2020-07-29 at 14 13 43@2x

Comment (rtibbles) [reply kolibri/plugins/coach/assets/src/modules/coachNotifications/getters.js:142]: Ah yep - I am filtering the ANSWERED events, but doing so after this step, so just moving it up would be the answer here. Good catch.

Comment (rtibbles) [reply kolibri/plugins/coach/assets/src/modules/coachNotifications/getters.js:142]: This should now be addressed - as long as Travis is fine, we should be good to merge.

**Review by jonboiser — APPROVED** [review]

**Pull request merged**